pb/logger_interceptor.go

```diff
@@ -0,0 +1,118 @@
+package pb
+
+import (
+	"context"
+	"path"
+	"time"
+
+	"google.golang.org/grpc"
+)
+
+// LogFn is the function used to log the messages
+type LogFn func(msg string, fields Fields)
```

Comment:
Maybe this should be `type LogFn func(format string, args ...interface{}, fields Fields)` to make it more generic, like go-log & logrus `Infof(format string, args ...interface{})`.

Also, maybe it would be good to include the interceptors in this file in the language-analyzer example.

Reply:
Probably the variadic arg needs to be last, so maybe it should be `type LogFn func(fields Fields, format string, args ...interface{})`

Reply:
I changed `LogFn`, but I've just noticed that if we want to have both `CtxlogStreamClientInterceptor` and `CtxlogUnaryClientInterceptor` as default dial options as discussed [here](https://github.com/src-d/lookout-sdk/pull/77#discussion_r259795368), then the user cannot pass any other interceptors. How do we want to handle this? Are we leaving them as non-default?

Reply:
Could you please point to the place that prevents the user to pass other interceptors?

Reply:
AFAIU it is possible to pass only one stream and one unary interceptor, that's why we need to use the chain functions provided by grpc middleware.

Reply:
oh. I see. Interceptors are wrapped in `DialOption` which is private and we can't build the chain in our constructor. I see only 2 solutions for this problem:
1. Change the signature of `DialContext` and make it super ugly
2. Change code from:
```go
	opts = append(opts,
		grpc.WithStreamInterceptor(CtxlogStreamClientInterceptor),
		grpc.WithUnaryInterceptor(CtxlogUnaryClientInterceptor),
```
to:
```go
	opts = append([]DialOption{},
		grpc.WithStreamInterceptor(CtxlogStreamClientInterceptor),
		grpc.WithUnaryInterceptor(CtxlogUnaryClientInterceptor),
                opts...)
```

so the user can override interceptors. And document necessity of including `*Ctxlog` in the chain if they do so. Which is not very cool too.

Reply:
I don't like relying on the user for including `Ctxlog*` interceptors.

I came up with another solution that is a bit verbose, but IMO is clean and doesn't make the `DialContext` signature ugly.
```
type DialOptions interface {
	Finalize() []grpc.DialOption
}

func NewDialOptions() DialOptions {
	dOpts := &dialOptions{
		streamClientInts: []grpc.StreamClientInterceptor{CtxlogStreamClientInterceptor},
		unaryClientInts:  []grpc.UnaryClientInterceptor{CtxlogUnaryClientInterceptor},
		insecure:         true,
		callOptions: []grpc.CallOption{
			grpc.MaxCallRecvMsgSize(maxMessageSize),
			grpc.MaxCallSendMsgSize(maxMessageSize),
		},
	}

	return dOpts
}

type dialOptions struct {
	streamClientInts []grpc.StreamClientInterceptor
	unaryClientInts  []grpc.UnaryClientInterceptor
	insecure         bool
	callOptions      []grpc.CallOption
}

func (d *dialOptions) WithStreamInterceptor(sci grpc.StreamClientInterceptor) *dialOptions {
	d.streamClientInts = append(d.streamClientInts, sci)
	return d
}

func (d *dialOptions) WithUnaryInterceptor(uci grpc.UnaryClientInterceptor) *dialOptions {
	d.unaryClientInts = append(d.unaryClientInts, uci)
	return d
}

func (d *dialOptions) WithInsecure() *dialOptions {
	d.insecure = true
	return d
}

func (d *dialOptions) WithDefaultCallOptions(cos ...grpc.CallOption) *dialOptions {
	d.callOptions = append(d.callOptions, cos...)
	return d
}

// ...other setters...

func (d *dialOptions) Finalize() []grpc.DialOption {
	var dialOptions []grpc.DialOption
	dialOptions = append(dialOptions,
		grpc.WithUnaryInterceptor(grpc_middleware.ChainUnaryClient(d.unaryClientInts...)))
	dialOptions = append(dialOptions,
		grpc.WithStreamInterceptor(grpc_middleware.ChainStreamClient(d.streamClientInts...)))
	dialOptions = append(dialOptions, grpc.WithInsecure())
	dialOptions = append(dialOptions, grpc.WithDefaultCallOptions(d.callOptions...))

	return dialOptions
}

// DialContext creates a client connection to the given target with custom message size
func DialContext(ctx context.Context, target string, dOpts DialOptions) (*grpc.ClientConn, error) {
	opts := dOpts.Finalize()
	return grpc.DialContext(ctx, target, opts...)
}
```

So the user can only build a `dialOptions` only using the factory `NewDialOptions`, and the factory is responsible for adding the defaults options. WDYT?

Reply:
You don't even need `NewDialOptions` you can mimic grpc api completely. Though in my opinion it's overkill solution for this problem.

Reply:
Ok, I don't have any preference then on the two options you proposed. Maybe the second one is slightly better.

Reply:
/cc @carlosms could we have your input on it?

Reply:
I agree with @smacker that I would prefer to avoid the overkill solution, if possible.

What about something like this? It's a bit ugly and we need to document the need to use `WithInterceptors`, but at least for the base use case the signature does not change, and we don't rely on the user remembering to add our `Ctxlog*` interceptors...

```go
func DialContext(ctx context.Context, target string, opts ...grpc.DialOption) (*grpc.ClientConn, error) {
	return DialContextWithInterceptors(ctx, target, nil, nil, opts...)
}

func DialContextWithInterceptors(
	ctx context.Context,
	target string,
	streamInterceptors []grpc.StreamClientInterceptor,
	unaryInterceptors []grpc.UnaryClientInterceptor,
	opts ...grpc.DialOption,
) (*grpc.ClientConn, error) {
	streamInterceptors = append(
		[]grpc.StreamClientInterceptor{CtxlogStreamClientInterceptor},
		streamInterceptors...)
	unaryInterceptors = append(
		[]grpc.UnaryClientInterceptor{CtxlogUnaryClientInterceptor},
		unaryInterceptors...)

	opts = append(opts,
		grpc.WithStreamInterceptor(grpc_middleware.ChainStreamClient(
			streamInterceptors...,
		)),
		grpc.WithUnaryInterceptor(grpc_middleware.ChainUnaryClient(
			unaryInterceptors...,
		)),
		grpc.WithDefaultCallOptions(
			grpc.MaxCallRecvMsgSize(maxMessageSize),
			grpc.MaxCallSendMsgSize(maxMessageSize),
		),
		grpc.WithInsecure(),
	)

	return grpc.DialContext(ctx, target, opts...)
}
```

To be used like this when you need to use extra interceptors:
```go
	conn, err := pb.DialContextWithInterceptors(ctx, dataSrvAddr,
		[]grpc.StreamClientInterceptor{pb.LogStreamClientInterceptor(logfn)},
		[]grpc.UnaryClientInterceptor{pb.LogUnaryClientInterceptor(logfn)},
		grpc.WithBlock(),
		grpc.WithInitialWindowSize(1234),
	)
```

Plus the equivalent for `NewServer` / `NewServerWithInterceptors`.

Reply:
I'm good either with just docs for the necessity to include our interceptors or with what Carlos proposed. It's a mix of 2 previous solutions that looks very resonable and gives more control to us than just docs (for example if we decide to add more default interceptors in the future).

Also tbh I think we are solving a quite rare case. I believe most people (and we ourselves) will just use default `DialContext` without any extra options.

Reply:
> Also tbh I think we are solving a quite rare case. I believe most people (and we ourselves) will just use default `DialContext` without any extra options.

Fair enough, but I like @carlosms solution, I think it's a very good compromise. And you both pointed out it's not very cool to rely on the user. 👍 thanks!

Reply:
Addressed [here](https://github.com/src-d/lookout-sdk/pull/77/commits/21c1af1cf837b3e99607e2c78f762d79ce138eb6)